deps/wildcard/wildcard.c

```diff
@@ -0,0 +1,169 @@
+#include <stdio.h>
+#include <assert.h>
+
+#include "wildcard.h"
+
+match_t Wildcard_MatchChar(const char *pattern, size_t p_len, const char *str, size_t str_len) {
+  const char *pattern_end = pattern + p_len;
+  const char *str_end = str + str_len;
+  const char *pattern_itr = pattern;
+  const char *str_itr = str;
+
+  const char *np_itr = NULL;
+  const char *ns_itr = NULL;
+  int i = 0;
+  while (++i) {
+    //printf("%d ", i);
+    if (pattern_end != pattern_itr) {
+      const char c = *pattern_itr;
+      if ((str_end != str_itr) && (c == *str_itr || c == '?')) {
+        ++str_itr;
+        ++pattern_itr;
+        continue;
+      } else if (c == '*') {
+        while ((pattern_end != pattern_itr) && (*pattern_itr == '*')) {
```

Comment:
`pattern_end` points to one char after the `pattern` buffer, so here you might access invalid memory (unless you assume its a C terminated strings and you have the `\0` ah the end, in such case why you pass the `len`).
General note, I believe that Redis already has a pattern matching function that is used on `keys` and `scan` command, why don't we take it?

Reply:
Also took that code to gears:
https://github.com/RedisGears/RedisGears/blob/7927cd41d560a810a54dcaf7170cc172355a4a15/src/readers/keys_reader.c#L558

Reply:
I will look into the other functions.

Reply:
redis is using the `glob` library. This is a good solution for `char` strings but not for `rune` strings.